core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1080,6 +1080,21 @@ abstract class RDD[T: ClassTag](
       var partiallyAggregated = mapPartitions(it => Iterator(aggregatePartition(it)))
       var numPartitions = partiallyAggregated.partitions.length
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
+
+      // Do one level of aggregation based on executorId before starting the tree
+      // NOTE: exclude the driver from list of executors
+      val numExecutors = math.max(context.getExecutorStorageStatus.length - 1, 1)
+      partiallyAggregated = partiallyAggregated.mapPartitionsWithIndex { case (idx, iter) =>
+        def isAllDigits(x: String) = x forall Character.isDigit
+        val execId = SparkEnv.get.executorId
+        if (isAllDigits(execId)) {
+          iter.map((execId.toInt, _))
+        } else {
+          iter.map((execId.hashCode, _))
```

Comment:
oh I see it's because `HashPartitioner` would hash that string anyway. One thing we could do is to just try to cast it to an int and if it doesn't work we fall back to the old approach. Unfortunately it's a little more code.